test/daemon/pubsub.spec.js

```diff
@@ -0,0 +1,295 @@
+      client = new Client(daemonAddr)
+
+      await client.attach()
+
+      // connect peers
+      let request = {
+        type: Request.Type.CONNECT,
+        connect: {
+          peer: Buffer.from(libp2pPeer.peerInfo.id.toBytes()),
+          addrs: libp2pPeer.peerInfo.multiaddrs.toArray().map(addr => addr.buffer)
+        },
+        streamOpen: null,
+        streamHandler: null,
+        dht: null,
+        disconnect: null,
+        pubsub: null,
+        connManager: null
+      }
+
+      let stream = client.send(request)
+
+      let message = await stream.first()
+      let response = Response.decode(message)
+      expect(response.type).to.eql(Response.Type.OK)
+      stream.end()
+
+      // subscribe topic
+      await libp2pPeer.pubsub.subscribe(topic, {}, (msg) => {
+        expect(msg.data).to.equalBytes(data)
+        resolve()
+      })
+
+      // wait to pubsub to propagate messages
+      await new Promise(resolve => setTimeout(resolve, 1000))
+
+      // publish topic
+      request = {
+        type: Request.Type.PUBSUB,
+        connect: null,
+        streamOpen: null,
+        streamHandler: null,
+        pubsub: {
+          type: PSRequest.Type.PUBLISH,
+          topic,
+          data: data
+        },
+        disconnect: null,
+        connManager: null
+      }
+
+      stream = client.send(request)
+
+      response = Response.decode(await stream.first())
+      expect(response.type).to.eql(Response.Type.OK)
+
+      stream.end()
+    })
+  })
+
+  it('should be able to receive messages from subscribed topics', function () {
+    this.timeout(20e3)
+
+    const topic = 'test-topic'
+    const data = Buffer.from('test-data')
+
+    return new Promise(async (resolve) => {
+      client = new Client(daemonAddr)
+
+      await client.attach()
+
+      // connect peers
+      let request = {
+        type: Request.Type.CONNECT,
+        connect: {
+          peer: Buffer.from(libp2pPeer.peerInfo.id.toBytes()),
+          addrs: libp2pPeer.peerInfo.multiaddrs.toArray().map(addr => addr.buffer)
+        }
+      }
+
+      let stream = client.send(request)
+
+      let message = await stream.first()
+      let response = Response.decode(message)
+      expect(response.type).to.eql(Response.Type.OK)
+      stream.end()
+
+      // subscribe topic
+      request = {
+        type: Request.Type.PUBSUB,
+        connect: null,
+        streamOpen: null,
+        streamHandler: null,
+        pubsub: {
+          type: PSRequest.Type.SUBSCRIBE,
+          topic
+        },
+        disconnect: null,
+        connManager: null
+      }
+
+      stream = client.send(request)
+
+      let subscribed = false
+
+      for await (const msg of stream) {
+        if (subscribed) {
+          response = PSMessage.decode(msg)
+          expect(response).to.exist()
+          expect(response.data).to.exist()
+          expect(response.data[0]).to.exist()
+          expect(response.data[0]).to.equalBytes(data)
```

Comment:
It would be good to check the latest pubsub fields. PBMessage has changed since I added the proto, so it may not work with the latest go-daemon. https://github.com/libp2p/go-libp2p-daemon/blob/v0.0.2/pb/p2pd.proto#L146. Data is no longer repeated, for example.  Here's the commit change for pb message, https://github.com/libp2p/go-libp2p-daemon/commit/95ab1d203345f93e2d8bcb1c17ae8dafcc758712#diff-d864c93b68284fafd809efbf29a1f0eb

Reply:
I will update them!